File: Ant/chapter3/cpp/4_values_whose_sum_is_0.cpp
#include <algorithm>
#include <fstream>
#include <stdio.h>

#define MAX_N 4000

using namespace std;

typedef long long ll;

int n;
ll A[MAX_N], B[MAX_N], C[MAX_N], D[MAX_N];
ll CD[MAX_N];

int main() {
  ifstream ifs("../testset/4_values_whose_sum_is_0/test1.txt");
  ifs >> n;
  for (int i = 0; i < n; ++i) {
    ifs >> A[i];
  }
  for (int i = 0; i < n; ++i) {
    ifs >> B[i];
  }
  for (int i = 0; i < n; ++i) {
    ifs >> C[i];
  }
  for (int i = 0; i < n; ++i) {
    ifs >> D[i];
  }
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < n; ++j) {
      CD[i * n + j] = C[i] + D[j];
    }
  }
  sort(CD, CD + n * n);
  ll ans = 0;
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < n; ++j) {
      ll cd = -(A[i] + B[j]);
      ans += upper_bound(CD, CD + n * n, cd) - lower_bound(CD, CD + n * n, cd); // num of CD[k] == cd
    }
  }
  printf("%lld\n", ans);
}

File: Ant/chapter3/python/4_values_whose_sum_is_0.py
from bisect import bisect_left, bisect


def ns(f):
    return next(f).strip()


with open("../testset/4_values_whose_sum_is_0/test1.txt", 'r') as f:
    n = int(ns(f))
    A = list(map(int, ns(f).split()))
    B = list(map(int, ns(f).split()))
    C = list(map(int, ns(f).split()))
    D = list(map(int, ns(f).split()))


CD = []
for c in C:
    for d in D:
        CD.append(c + d)
CD.sort()
ans = 0
for a in A:
    for b in B:
        cd = -(a + b)
        ans += bisect(CD, cd) - bisect_left(CD, cd)
print(ans)
